### DISCUSSION TITLE
Middleware alternative with App Router?

### DISCUSSION THREAD

**Discussion by levipadre** [amannn/next-intl#291] [Q&A]
I started to work on a website (App Router), and I have multiple domains, like domain.fr and domain.es and domain.org.uk etc.
The domains have different page structures, so it's not a mirror website.
I have a middleware with the following code:

```
import createMiddleware from 'next-intl/middleware';
 
export default createMiddleware({
    locales: ['en', 'fr'],
    defaultLocale: 'en',
    alternateLinks: false,
    domains: [
        {
            domain: 'next.domain.com',
            defaultLocale: 'en',
            locales: ['en']
        },
        {
            domain: 'next.domain.fr',
            defaultLocale: 'fr',
            locales: ['fr']
        }
    ],
    localeDetection: false
});

export const config = {
    matcher: ['/((?!api|_next|.*\\..*).*)']
}
```
I'm using [lang] inside the app folder, and everything is working well, but I would like to know if it's possible to handle this simple thing some other ways, like **vercel.json** or **next.config.js**. The reason: I have 12 domain names with hundreds of pages/domain, around 10,000 users per month/domain, so using middleware, I would reach the 1M limit quickly.
I asked on Next.js discussion page as well, but maybe someone can help me here too.
Code examples would be welcome.

**Comment by amannn** [comment]
Do I understand it correctly, that you're trying to find a solution to reduce the number of middleware invocations?

In the RSC beta, you currently have to call the middleware on all content routes. Therefore it's unfortunately not possible to restrict the `matcher` further.

However, if you use [Client Components (stable)](https://next-intl-docs.vercel.app/docs/next-13/client-components), you can theoretically limit the middleware to `/`. This however has the effect that the cookie that remembers the last matched locale is not updated for routes other than `/`. Since you have `localeDetection` turned off, this might not be an issue for you though.

I think the RSC integration will have the same option in the future, but I'm waiting for [`createServerContext`](https://react.dev/blog/2023/05/03/react-canaries#announcing-breaking-changes-and-new-features-early) to enable this.

So I guess in your current situation, you have to pick a tradeoff that works for you and the project. Hope this helps!
